### PR TITLE
3Cs base

### PR DESCRIPTION
**Pull request by afourneaux**
- Character: An upright version of the Portapotty model is set up as a prefab. It should be drag-and-droppable to any scene, but for ease of testing a scene with a simple flat grass plane also exists
- Camera: The camera locks to the player at a distance of 40 units, and the arrow keys can be used to rotate the camera around in a 3rd person circle. When the player is moving, the camera smoothly drags backwards to point in the direction the rocket is facing. The pull speed is modified by the speed of the rocket and how far it is from its final correct angle.
- Controls: Press SPACE for ignition to start the ship. Use WSAD to rotate the ship. Rotations are all relative to the camera. The intention is to feel chaotic and hard to control, but still wild and fun.

### CODE REVIEW

**Review by Aether700 — CHANGES_REQUESTED** [review]
Holding one of the arrow keys down leads to the camera flickering and spasing out. Could you please fix that?
To recreate press spacebar then hold down any of the arrow keys.

**Comment by afourneaux** [comment]
Issue resolved, camera is now smooth at all times

File: Assets/Scripts/CameraController.cs [hunk 1]
```diff
@@ -33,12 +33,11 @@ void LateUpdate()
         Camera.main.transform.Rotate(new Vector3(0,1,0), horizontal * cameraSpeed * dt);
         Camera.main.transform.Rotate(new Vector3(1,0,0), vertical * cameraSpeed * dt);
```
Comment (Aether700) [inline Assets/Scripts/CameraController.cs]: Please do no use Camera.main to perform multiple operations as it is a very costly function. Instead use it to get a reference to the main camera in the Start/Awake function and use the reference when doing the operations

**Review by Aether700 — COMMENTED** [review]
Much better. There doesn't seem to be a way to rotate the camera around the z axis of the rocket which makes understanding what is going on very hard. I understand that the controls are meant to be hard to use but right now I can barely tell what's going on because I can barely see.

**Review by Aether700 — DISMISSED** [review]

File: Assets/Scripts/CameraController.cs (hunk 1 above)
Comment (afourneaux) [reply Assets/Scripts/CameraController.cs]: TIL. Here I assumed it was a cached reference in the Camera class rather than a FindObjectWithTag wrapper

**Comment by afourneaux** [comment]
> Much better. There doesn't seem to be a way to rotate the camera around the z axis of the rocket which makes understanding what is going on very hard. I understand that the controls are meant to be hard to use but right now I can barely tell what's going on because I can barely see.

You're looking for a roll for the camera? I can certainly do, but considering the camera uses the arrow keys do you have an idea for third axis controls for that?

**Comment by Aether700** [comment]
> > Much better. There doesn't seem to be a way to rotate the camera around the z axis of the rocket which makes understanding what is going on very hard. I understand that the controls are meant to be hard to use but right now I can barely tell what's going on because I can barely see.
> 
> You're looking for a roll for the camera? I can certainly do, but considering the camera uses the arrow keys do you have an idea for third axis controls for that?

no I don't sadly. I don't but given that we want to use a controller in the long run maybe try maping those to the controller? for example, use the joystick to rotate like you have been rotating with the arrow keys and have the triggers rotate on the forward axis.

As of now you would have to remap the rotation to something like HJKU and add Y and I to rotate around the forward axis because using the arrow keys would not allow for an easy use of other keys to rotate around that additional axis.

Another possible solution would be to fix the rotation of the camera on the forward axis with the model so rotating the model would rotate the camera around that axis also.

In any case idk which one of those you would like to do but they should give you a good head start on fixing that issue.

**Comment by afourneaux** [comment]
> > > Much better. There doesn't seem to be a way to rotate the camera around the z axis of the rocket which makes understanding what is going on very hard. I understand that the controls are meant to be hard to use but right now I can barely tell what's going on because I can barely see.
> > 
> > 
> > You're looking for a roll for the camera? I can certainly do, but considering the camera uses the arrow keys do you have an idea for third axis controls for that?
> 
> no I don't sadly. I don't but given that we want to use a controller in the long run maybe try maping those to the controller? for example, use the joystick to rotate like you have been rotating with the arrow keys and have the triggers rotate on the forward axis.
> 
> As of now you would have to remap the rotation to something like HJKU and add Y and I to rotate around the forward axis because using the arrow keys would not allow for an easy use of other keys to rotate around that additional axis.
> 
> Another possible solution would be to fix the rotation of the camera on the forward axis with the model so rotating the model would rotate the camera around that axis also.
> 
> In any case idk which one of those you would like to do but they should give you a good head start on fixing that issue.

I was actually mostly concerned with this issue because of theoretical eventual controller controls, joysticks can't represent roll movement and often use shoulder buttons in the rare case they're necessary. Fixing the rotation wouldn't likely be ideal, since the user can and will achieve Z rotation just by virtue of a lot of XY rotation. I'll assign HJKUIY for now and have a ponder on how to do it more gracefully in the long term, or if it's necessary to do at all.

**Comment by Aether700** [comment]
> 

you could lock that rotation axis but make sure to have the "rocket nose" parallel to the xz plane because as of right now it is parallel to it. idk how complex you want the controls to be but I personally would just lock that rotation of the camera and make sure it stays constant

**Review by Aether700 — CHANGES_REQUESTED** [review]
Looks good.

Two small details would be that I would invert the rotation of Y and I and that we might want to rework the controls for the final game but for a first pass it's great.

**Review by Aether700 — DISMISSED** [review]

**Review by Aether700 — APPROVED** [review]

**Pull request merged**